distributed/worker.py

```diff
@@ -2016,6 +2019,17 @@ def handle_compute_task(
             for key, value in nbytes.items():
                 self.tasks[key].nbytes = value
 
+    def _add_to_data_needed(self, ts: TaskState, stimulus_id: str) -> RecsInstrs:
+        self.data_needed.push(ts)
+        for w in ts.who_has:
+            self.data_needed_per_worker[w].push(ts)
+        # This is the same as `return self._ensure_communicating()`, except that when
+        # many tasks transition to fetch at the same time, e.g. from a single
+        # compute-task or acquire-replicas command from the scheduler, it allows
+        # clustering the transfers into less GatherDep instructions; see
+        # _select_keys_for_gather().
+        return {}, [EnsureCommunicatingLater(stimulus_id=stimulus_id)]
```

Comment:
The alternative to this was to delete ``_select_keys_for_gather`` and either

1. Add logic to ``_handle_instructions`` to squash individual GatherDep instructions on the same worker
2. Implement no grouping whatsoever and just rely on rpc pooling (needs performance testing)

Both systems would remove the out-of-priority fetch from workers and imply revisiting the limits for concurrent connections.

Either way it would be a major functional change so I opted for this somewhat dirty hack instead which is functionally identical to main.